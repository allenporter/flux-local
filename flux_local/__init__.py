"""
.. include:: ../README.md
"""

__all__ = [
    "git_repo",
    "manifest",
    "kustomize",
    "helm",
    "exceptions",
]
