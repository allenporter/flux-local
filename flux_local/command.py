"""Library for issuing commands using asyncio and returning the result."""

import asyncio
from abc import ABC, abstractmethod
import logging
import shlex
import subprocess
from dataclasses import dataclass
from pathlib import Path
from collections.abc import Sequence
import os

from .exceptions import CommandException

_LOGGER = logging.getLogger(__name__)

_CONCURRENCY = 20
_SEM = asyncio.Semaphore(_CONCURRENCY)
_TIMEOUT = 20.0


# No public API
__all__: list[str] = []


class Task(ABC):
    """An instance of a async task to execute."""

    @abstractmethod
    async def run(self, stdin: bytes | None = None) -> bytes:
        """Execute the task and return the result."""


def format_path(path: Path) -> str:
    """Format path for debugging."""
    if path.is_absolute():
        cwd = Path.cwd()
        if path.is_relative_to(cwd):
            rel_path = str(path.relative_to(cwd))
            return f"{rel_path} (abs)"
    return str(path)


@dataclass
class Command(Task):
    """An instance of a command to run."""

    cmd: list[str]
    """Array of command line arguments."""

    cwd: Path | None = None
    """Current working directory."""

    exc: type[CommandException] = CommandException
    """Exception to throw in case of an error."""

    retcodes: list[int] | None = None
    """Non-zero error codes that are allowed to indicate success (e.g. for diff)."""

    env: dict[str, str] | None = None
    """Environment variables for the subprocess."""

    @property
    def string(self) -> str:
        """Render the command as a single string."""
        return " ".join([shlex.quote(arg) for arg in self.cmd])

    def __str__(self) -> str:
        """Render as a debug string."""
        cwd: str = ""
        if self.cwd:
            cwd = f"({format_path(self.cwd)}) "
        return f"{cwd}{self.string}"

    async def run(self, stdin: bytes | None = None) -> bytes:
        """Run the command, returning stdout."""
        _LOGGER.debug("Running command: %s", self)
        env = {
            **os.environ,
            **(self.env if self.env else {}),
        }
        proc = await asyncio.create_subprocess_shell(
            self.string,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            cwd=self.cwd,
            env=env,
        )
        out, err = await proc.communicate(stdin)
        if proc.returncode:
            if self.retcodes and proc.returncode in self.retcodes:
                return out
            errors = [f"Command '{self}' failed with return code {proc.returncode}"]
            if out:
                errors.append(out.decode("utf-8"))
            if err:
                errors.append(err.decode("utf-8"))
            _LOGGER.debug("\n".join(errors))
            raise self.exc("\n".join(errors))
        return out


async def _run_piped_with_sem(cmds: Sequence[Task]) -> str:
    """Run a set of commands, piped together, returning stdout of last."""
    stdin = None
    out = None
    for cmd in cmds:
        try:
            out = await asyncio.wait_for(cmd.run(stdin), _TIMEOUT)
        except asyncio.exceptions.TimeoutError as err:
            raise cmd.exc(f"Command '{cmd}' timed out") from err
        stdin = out
    return out.decode("utf-8") if out else ""


async def run_piped(cmds: Sequence[Task]) -> str:
    """Run a set of commands, piped together, returning stdout of last."""
    async with _SEM:
        result = await _run_piped_with_sem(cmds)
    return result


async def run(cmd: Task) -> str:
    """Run the specified command and return stdout."""
    return await run_piped([cmd])
