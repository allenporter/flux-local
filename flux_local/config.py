"""Configuration objects for flux-local."""

from dataclasses import dataclass


@dataclass
class HelmControllerConfig:
    """Configuration for the HelmReleaseController."""

    wipe_secrets: bool = True


@dataclass
class KustomizationControllerConfig:
    """Configuration for the KustomizationController."""

    wipe_secrets: bool = True


@dataclass
class ReadAction:
    """Configuration for reading resources."""

    wipe_secrets: bool = True
