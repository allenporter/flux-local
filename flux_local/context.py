"""Utilities for context tracing."""

import contextvars
from contextlib import contextmanager
import logging
from time import perf_counter
from typing import Generator


_LOGGER = logging.getLogger(__name__)

# No public API
__all__: list[str] = []


trace: contextvars.ContextVar[list[str]] = contextvars.ContextVar("trace")


@contextmanager
def trace_context(name: str) -> Generator[None, None, None]:
    stack = trace.get([])
    token = trace.set(stack + [name])
    label = " > ".join(stack + [name])
    t1 = perf_counter()
    _LOGGER.debug("[Trace] > %s", label)
    try:
        yield
    finally:
        t2 = perf_counter()
        trace.reset(token)
        _LOGGER.debug("[Trace] < %s (%0.2fs)", label, (t2 - t1))
