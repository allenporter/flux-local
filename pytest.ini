[pytest]
log_level = DEBUG
asyncio_mode = auto
