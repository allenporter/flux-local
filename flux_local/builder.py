"""Library for building cluster resources into a Manifest.

This will build a `manifest.Manifest` from a cluster repo. This follows the
pattern of building kustomizations, then reading helm releases (though it will
not evaluate the templates). The resulting `Manifest` contains all the necessary
information to do basic checks on objects in the cluster (e.g. run templates
from unit tests).

Example usage:
```
from flux_local import builder

manifest = await builder.manifest()
for cluster in manifest:
    print(f"Found cluster: {cluster.path}")
    for kustomization in cluster.kustomizations:
        print(f"Found kustomization: {kustomization.path}")
```
"""

import logging
from collections.abc import Callable
from pathlib import Path
from typing import Any

from . import kustomize
from .manifest import (
    CLUSTER_KUSTOMIZE_DOMAIN,
    KUSTOMIZE_DOMAIN,
    Cluster,
    HelmRelease,
    HelmRepository,
    Kustomization,
    Manifest,
)

__all__ = [
    "manifest",
]

_LOGGER = logging.getLogger(__name__)

CLUSTER_KUSTOMIZE_NAME = "flux-system"
CLUSTER_KUSTOMIZE_KIND = "Kustomization"
KUSTOMIZE_KIND = "Kustomization"
HELM_REPO_KIND = "HelmRepository"
HELM_RELEASE_KIND = "HelmRelease"


def domain_filter(version: str) -> Callable[[dict[str, Any]], bool]:
    """Return a yaml doc filter for specified resource version."""

    def func(doc: dict[str, Any]) -> bool:
        if api_version := doc.get("apiVersion"):
            if api_version.startswith(version):
                return True
        return False

    return func


CLUSTER_KUSTOMIZE_DOMAIN_FILTER = domain_filter(CLUSTER_KUSTOMIZE_DOMAIN)
KUSTOMIZE_DOMAIN_FILTER = domain_filter(KUSTOMIZE_DOMAIN)


async def get_clusters(path: Path) -> list[Cluster]:
    """Load Cluster objects from the specified path."""
    cmd = kustomize.grep(f"kind={CLUSTER_KUSTOMIZE_KIND}", path).grep(
        f"metadata.name={CLUSTER_KUSTOMIZE_NAME}"
    )
    docs = await cmd.objects()
    return [
        Cluster.from_doc(doc) for doc in docs if CLUSTER_KUSTOMIZE_DOMAIN_FILTER(doc)
    ]


async def get_kustomizations(path: Path) -> list[Kustomization]:
    """Load Kustomization objects from the specified path."""
    cmd = kustomize.grep(f"kind={KUSTOMIZE_KIND}", path).grep(
        f"metadata.name={CLUSTER_KUSTOMIZE_NAME}",
        invert=True,
    )
    docs = await cmd.objects()
    return [Kustomization.from_doc(doc) for doc in docs if KUSTOMIZE_DOMAIN_FILTER(doc)]


async def manifest(root: Path) -> Manifest:
    """Build a Manifest object from the local cluster.

    This will locate all Kustomizations that represent clusters, then find all
    the Kustomizations within that cluster, as well as all relevant Helm
    resources.
    """

    clusters = await get_clusters(root)
    for cluster in clusters:
        _LOGGER.debug("Processing cluster: %s", cluster.path)
        cluster.kustomizations = await get_kustomizations(
            root / cluster.path.lstrip("./")
        )
        for kustomization in cluster.kustomizations:
            _LOGGER.debug("Processing kustomization: %s", kustomization.path)
            cmd = kustomize.build(root / kustomization.path)
            kustomization.helm_repos = [
                HelmRepository.from_doc(doc)
                for doc in await cmd.grep(f"kind=^{HELM_REPO_KIND}$").objects()
            ]
            kustomization.helm_releases = [
                HelmRelease.from_doc(doc)
                for doc in await cmd.grep(f"kind=^{HELM_RELEASE_KIND}$").objects()
            ]
    return Manifest(clusters=clusters)
