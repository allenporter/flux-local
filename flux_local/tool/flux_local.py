"""Command line tool for building, diffing, validating flux local repositories."""

import argparse
import asyncio
import logging
import pathlib
import sys

import yaml

from . import build, diff, get, manifest, test
from flux_local import command

_LOGGER = logging.getLogger(__name__)


def main() -> None:
    """Flux-local command line tool main entry point."""
    # https://github.com/yaml/pyyaml/issues/89
    yaml.Loader.yaml_implicit_resolvers.pop("=")

    parser = argparse.ArgumentParser(
        description="Command line utility for inspecting a local flux repository.",
    )
    parser.add_argument(
        "--log-level", choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]
    )

    subparsers = parser.add_subparsers(dest="command", help="Command", required=True)

    build_args = subparsers.add_parser(
        "build",
        help="Build local flux Kustomization target from a local directory",
    )
    build_args.add_argument(
        "path", type=pathlib.Path, help="Path to the kustomization or charts"
    )
    build_args.add_argument(
        "--enable-helm",
        type=bool,
        action=argparse.BooleanOptionalAction,
        help="Enable use of HelmRelease inflation",
    )
    # pylint: disable=duplicate-code
    build_args.add_argument(
        "--skip-crds",
        type=bool,
        default=False,
        action=argparse.BooleanOptionalAction,
        help="Allows disabling of outputting CRDs to reduce output size",
    )
    build_args.set_defaults(cls=build.BuildAction)

    manifest_args = subparsers.add_parser(
        "manifest", help="Build a yaml manifest file representing the cluster"
    )
    manifest_args.add_argument(
        "path", type=pathlib.Path, help="Path to the kustomization or charts"
    )
    manifest_args.set_defaults(cls=manifest.ManifestAction)

    test_args = subparsers.add_parser("test", help="Build and validate the cluster")
    test_args.add_argument(
        "path", type=pathlib.Path, help="Path to the kustomization or charts"
    )
    test_args.set_defaults(cls=test.TestAction)

    get.GetAction.register(subparsers)
    diff.DiffAction.register(subparsers)

    args = parser.parse_args()

    if args.log_level:
        logging.basicConfig(level=args.log_level)

    action = args.cls()
    try:
        asyncio.run(action.run(**vars(args)))
    except command.CommandException as err:
        print("Command failed: ", err)
        sys.exit(1)


if __name__ == "__main__":
    main()
