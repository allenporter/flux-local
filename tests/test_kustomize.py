"""Tests for kustomize library."""

from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion

from flux_local import kustomize, exceptions

TESTDATA_DIR = Path("tests/testdata")

KUSTOMIZATION = """---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- example.yaml
"""


@pytest.mark.parametrize(
    "path",
    [TESTDATA_DIR / "repo", (TESTDATA_DIR / "repo").absolute()],
)
async def test_grep(path: Path, snapshot: SnapshotAssertion) -> None:
    """Test a kustomize grep command."""
    result = await kustomize.grep("kind=ConfigMap", path).run()
    assert "Secret" not in result
    assert "ConfigMap" in result
    assert result == snapshot


@pytest.mark.parametrize(
    "path",
    [TESTDATA_DIR / "repo", (TESTDATA_DIR / "repo").absolute()],
)
async def test_objects(path: Path, snapshot: SnapshotAssertion) -> None:
    """Test loading yaml documents."""
    cmd = kustomize.grep("kind=ConfigMap", path)
    result = await cmd.objects()
    assert len(result) == 1
    assert result[0].get("kind") == "ConfigMap"
    assert result[0].get("apiVersion") == "v1"
    assert result == snapshot


@pytest.mark.parametrize(
    "path",
    [TESTDATA_DIR / "repo", (TESTDATA_DIR / "repo").absolute()],
)
async def test_stash(path: Path) -> None:
    """Test loading yaml documents."""
    cmd = await kustomize.grep("kind=Ignored", path, invert=True).stash()
    result = await cmd.grep("kind=ConfigMap").objects()
    assert len(result) == 1
    assert result[0].get("kind") == "ConfigMap"
    assert result[0].get("apiVersion") == "v1"
    result = await cmd.grep("kind=Secret").objects()
    assert len(result) == 1
    assert result[0].get("kind") == "Secret"
    assert result[0].get("apiVersion") == "v1"
    result = await cmd.grep("kind=Unknown").objects()
    assert len(result) == 0


@pytest.mark.parametrize(
    "path",
    [TESTDATA_DIR / "repo", (TESTDATA_DIR / "repo").absolute()],
)
async def test_validate_pass(path: Path) -> None:
    """Test applying policies to validate resources."""
    cmd = kustomize.grep("kind=ConfigMap", path)
    await cmd.validate(TESTDATA_DIR / "policies/pass.yaml")


@pytest.mark.parametrize(
    "path",
    [TESTDATA_DIR / "repo", (TESTDATA_DIR / "repo").absolute()],
)
async def test_validate_fail(path: Path) -> None:
    """Test applying policies to validate resources."""
    cmd = kustomize.grep("kind=ConfigMap", path)
    with pytest.raises(
        exceptions.CommandException, match="require-test-annotation: validation error"
    ):
        await cmd.validate(TESTDATA_DIR / "policies/fail.yaml")


async def test_target_namespace() -> None:
    """Test a kustomization with a target namespace."""
    ks = kustomize.grep("kind=ConfigMap", TESTDATA_DIR / "repo")

    result = await ks.objects()
    assert len(result) == 1
    config_map = result[0]
    assert "metadata" in config_map
    assert config_map["metadata"] == {
        "name": "cluster-settings",
        "namespace": "flux-system",
        "annotations": {
            "config.kubernetes.io/index": "0",
            "config.kubernetes.io/path": "cluster-settings.yaml",
            "internal.config.kubernetes.io/index": "0",
            "internal.config.kubernetes.io/path": "cluster-settings.yaml",
        },
    }

    result = await ks.objects(target_namespace="configs")
    assert len(result) == 1
    config_map = result[0]
    assert "metadata" in config_map
    assert config_map["metadata"] == {
        "name": "cluster-settings",
        # Verify updated namespace
        "namespace": "configs",
        "annotations": {
            "config.kubernetes.io/index": "0",
            "config.kubernetes.io/path": "cluster-settings.yaml",
            "internal.config.kubernetes.io/index": "0",
            "internal.config.kubernetes.io/path": "cluster-settings.yaml",
        },
    }
