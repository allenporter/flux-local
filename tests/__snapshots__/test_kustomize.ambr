# serializer version: 1
# name: test_grep[path0]
  '''
  apiVersion: v1
  kind: ConfigMap
  metadata:
    namespace: flux-system
    name: cluster-settings
    annotations:
      config.kubernetes.io/index: '0'
      config.kubernetes.io/path: 'cluster-settings.yaml'
      internal.config.kubernetes.io/index: '0'
      internal.config.kubernetes.io/path: 'cluster-settings.yaml'
  data:
    CLUSTER: dev
    DOMAIN: example.org
  
  '''
# ---
# name: test_grep[path1]
  '''
  apiVersion: v1
  kind: ConfigMap
  metadata:
    namespace: flux-system
    name: cluster-settings
    annotations:
      config.kubernetes.io/index: '0'
      config.kubernetes.io/path: 'cluster-settings.yaml'
      internal.config.kubernetes.io/index: '0'
      internal.config.kubernetes.io/path: 'cluster-settings.yaml'
  data:
    CLUSTER: dev
    DOMAIN: example.org
  
  '''
# ---
# name: test_objects[path0]
  list([
    dict({
      'apiVersion': 'v1',
      'data': dict({
        'CLUSTER': 'dev',
        'DOMAIN': 'example.org',
      }),
      'kind': 'ConfigMap',
      'metadata': dict({
        'annotations': dict({
          'config.kubernetes.io/index': '0',
          'config.kubernetes.io/path': 'cluster-settings.yaml',
          'internal.config.kubernetes.io/index': '0',
          'internal.config.kubernetes.io/path': 'cluster-settings.yaml',
        }),
        'name': 'cluster-settings',
        'namespace': 'flux-system',
      }),
    }),
  ])
# ---
# name: test_objects[path1]
  list([
    dict({
      'apiVersion': 'v1',
      'data': dict({
        'CLUSTER': 'dev',
        'DOMAIN': 'example.org',
      }),
      'kind': 'ConfigMap',
      'metadata': dict({
        'annotations': dict({
          'config.kubernetes.io/index': '0',
          'config.kubernetes.io/path': 'cluster-settings.yaml',
          'internal.config.kubernetes.io/index': '0',
          'internal.config.kubernetes.io/path': 'cluster-settings.yaml',
        }),
        'name': 'cluster-settings',
        'namespace': 'flux-system',
      }),
    }),
  ])
# ---
